base/src/main/scala/data/ProCompose.scala

```diff
@@ -0,0 +1,54 @@
+package scalaz
+package data 
+
+object ProCompose {
+  import typeclass.{Profunctor, Strong, Category, StrongClass, CategoryClass}
+  import typeclass.Strong._
+  import typeclass.Profunctor._
+
+  trait ProCompose[P[_, _], A, B] { } 
+
+  case class Pure[P[_, _], A, B](pab: A => B) extends ProCompose[P, A, B]
+
+  case class Composed[P[_, _], A, X, B](pab: P[A, X], more: ProCompose[P, X, B]) extends ProCompose[P, A, B]
```

Comment:
`Semiarrow` is just what one person called this construct on the Haskell mailing list; I don't know of a name that's been stable for a while for this, nor a few others in the `Semigroupoid->Strong/Choice` hierarchy.

My mistake, I misread; balanced composition sounds possible to me. My last concern is this: you need to pay a `O(n)` cost every time you compose a `ProCompose[F, A, B]` with a `ProCompose[F, B, C]` to perform the associated list concatenation, because of the cons-list structure. That could be fixed with a nice catenable `Queue` but we've no such thing yet; so `Vector` may be worth considering.